### DISCUSSION TITLE
Repo settings merge commit description and feature suggestion

### DISCUSSION THREAD

**Discussion by RobinKnipe** [community/community#51510] [Repositories]
### Select Topic Area

Product Feedback

## Content clarification
I think the description of the "**Allow merge commits**" option is wrong. In the repo's Settings > General > Pull Requests section:
<img width="799" alt="Screenshot 2023-03-30 at 16 57 10" src="https://user-images.githubusercontent.com/7834222/228909668-fa35b2af-5031-4820-936a-9a0eca12de08.png">
Surely merge commits are for bringing changes made to the base branch into the head branch? So the content should be more like:
> Add all commits from the **base** branch to the **head** branch with a merge commit.

## Feature suggestion
There is a risk of problems occurring when both merge commits _and_ linear history has been required (as part of branch protection rules). Does having both these settings enabled make any sense? Perhaps the settings UI could be improved to disable the "**Require linear history**" option (with an explanation message) when merge commits are enabled?

**Comment by kmk3** [comment]
> I think the description of the "**Allow merge commits**" option is wrong. In
> the repo's Settings > General > Pull Requests section:

> Surely merge commits are for bringing changes made to the base branch into
> the head branch? So the content should be more like:
> 
> > Add all commits from the **base** branch to the **head** branch with a
> > merge commit.

When comparing branches/commits in /compare (such as when trying to open a new
pull request), the following options are shown:

* base: [branch1] <- compare: [branch2]

This makes sense when comparing branches, as this effectively does
`git log "$branch1...$branch2"`.

That is, "base" here means "merge base" and `git merge-base "$base" HEAD` is in
most cases usually equivalent to `$base` or to a parent commit of it (see
`git-merge-base(1)`).  The GitHub documentation also uses "base" in this way:

* <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#three-dot-and-two-dot-git-diff-comparisons>

I think that the logic used in the settings is the following:

* "base branch" == merge base
* "head branch" == HEAD -> currently checked out topic branch

Which doesn't hold up too well in the context of merges, as when using
`git merge`, the branch that receives the merge is the branch that is currently
checked out (HEAD).

So I'd suggest to just name them **source** and **destination** in the settings
to make their descriptions as obvious as possible:

> Allow merge commits
> 
> Add all commits from the source branch into the destination branch with a
> merge commit.
> 
> Allow squash merging
> 
> Combine all commits from the source branch into a single commit in the
> destination branch.
> 
> Allow rebase merging
> 
> Copy each commit from the source branch onto the destination branch
> individually.

And to leave **base** and **compare** as is in /compare.
